From bf3e20f364e9bffd5cafca175efab8b3d194bb31 Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Wed, 26 Aug 2020 01:09:40 +0300
Subject: [PATCH] [SW-14676] fit: add start address

Add support for binary images which has a different start address than
the load address (begin of binary).

This change is needed in order to run Zephyr image which has it's start
address (entry point) in somewhere along the binary (i.e. not at the
beginning)

Change-Id: I8a1f2abf509b99eaff17b35f4db405c2d409030b
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/gaudi/common.c b/gaudi/common.c
index 3697522..3509ccd 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -260,7 +260,7 @@
 		if (strcmp(image[i].arch, "arm64"))
 			continue;
 
-		image_entry = (void *)image[i].loadaddr;
+		image_entry = (void *)image[i].startaddr;
 		break;
 	}
 
diff --git a/preboot/fit.c b/preboot/fit.c
index 4782c4a..90bca49 100644
--- a/preboot/fit.c
+++ b/preboot/fit.c
@@ -21,6 +21,7 @@
 #define FIT_HASHED_NODES_PROP	"hashed-nodes"
 #define FIT_DATA_PROP		"data"
 #define FIT_LOADADDR_PROP	"loadaddr"
+#define FIT_STARTADDR_PROP	"startaddr"
 #define FIT_ARCH_PROP		"arch"
 
 #define FIT_SEC_VERSION		"hl-sec-version"
@@ -456,13 +457,27 @@
 	return fit_verify_config(fit, noffset);
 }
 
+static int fit_getprop_u64(const void *fit, int node, const char *name,
+			   u64 *val64)
+{
+	const u32 *val32;
+	int len;
+
+	val32 = fdt_getprop(fit, node, name, &len);
+	if (!val32 || len != sizeof(u64))
+		return 1;
+
+	*val64 = ((u64)fdt32_to_cpu(val32[0])) << 32;
+	*val64 |= fdt32_to_cpu(val32[1]);
+	return 0;
+}
+
 enum fit_load_result fit_load_images(const void *fit, struct fit_image *images)
 {
 	int node, len;
 	int conf, num_images, i;
 	const u8 *buf;
-	const u32 *loadaddr;
-	u64 dst = 0;
+	u64 load, start;
 
 	/* Get number of images in preboot configuration */
 	conf = fit_find_config_node(fit);
@@ -486,14 +501,12 @@
 			return FIT_LOAD_CORRUPTED;
 		}
 
-		loadaddr = fdt_getprop(fit, node, FIT_LOADADDR_PROP, &len);
-		if (!loadaddr || len != sizeof(u64)) {
+		if (fit_getprop_u64(fit, node, FIT_LOADADDR_PROP, &load)) {
 			preboot_puts_hex("Wrong FIT format: loadaddr not found for image: ", i, false);
 			return FIT_LOAD_CORRUPTED;
 		}
-
-		dst = ((u64)fdt32_to_cpu(loadaddr[0])) << 32;
-		dst |= fdt32_to_cpu(loadaddr[1]);
+		if (fit_getprop_u64(fit, node, FIT_STARTADDR_PROP, &start))
+			start = load;
 
 		buf = fdt_getprop(fit, node, FIT_DATA_PROP, &len);
 		if (!buf) {
@@ -508,9 +521,10 @@
 			return FIT_LOAD_CORRUPTED;
 		}
 
-		images[i].loadaddr = dst;
+		images[i].loadaddr = load;
+		images[i].startaddr = start;
 
-		fit_copy_image(dst, buf, len);
+		fit_copy_image(load, buf, len);
 	}
 
 	return FIT_LOAD_SUCCESS;
diff --git a/preboot/fit.h b/preboot/fit.h
index dda3787..272036c 100644
--- a/preboot/fit.h
+++ b/preboot/fit.h
@@ -12,6 +12,7 @@
 
 struct fit_image {
 	u64 loadaddr;
+	u64 startaddr;
 	const char *arch;
 };
 
